File: solution/python/enrichment/bad_domain.py
import enricher
import logging
from generated import enrichment_pb2 as enpb
from generated import signal_pb2 as spb
from typing import Dict


INDICATOR_PATH = "../../data/indicators/bad_domain.csv"
COLUMN_NUM     = 2

def load_indicators() -> Dict[str, enpb.Host.Platform]:
    hostToOS = {} # dict[string]enpb.Host_Platform
    # Open indicators
    try:
        lst = open(INDICATOR_PATH)
    except Exception as e:
            logging.error("failed to open file %s: %s", INDICATOR_PATH, str(e))
            raise e

    for i in lst:
        strs = i.strip().split(",")
        hostToOS[strs[0]] = enpb.Host.Platform.Value(strs[1])

    lst.close()
    return hostToOS

class BadDomainEnricher(enricher.EnricherInterface):
    def __init__(self, name="", path=""):
        self.signals = []
        self.name = name
        self.signal_path = path

    # Get enricher name.
    def getName(self) -> str:
        return self.name
    
    # Get signals associated with a given enricher.
    def getSignals(self) -> list[spb.Signal]:
        return self.signals
    
    # Set the signals to be enriched.
    def setSignals(self, signals=[]) -> None:
        self.signals = signals
    
    # Enrich signals based on the provided client.
    def enrich(self, client) -> None:
        enrichedSigs = []
        hostOSDict = load_indicators()
        # For each signal, call our IPToHost enrichment based on source IP address.
        for sig in self.signals:
            try:
                resp = client.IPToHost(enpb.IP(
                    ip=sig.bad_domain.source_ip
                ))
                if resp.platform != hostOSDict[sig.bad_domain.bad_domain]:
                    logging.warn("domain indicator %s observed from %s - an operating system that isn't impacted", sig.bad_domain.bad_domain, enpb.Host.Platform.Name(resp.platform))
                else:
                    # Assign enriched hostname to protobuf field.
                    sig.bad_domain.hostname = resp.name
                    enrichedSigs.append(sig)
            except Exception as e:
                logging.warn("failed IPToHost conversion, output signals for manual investigation: %s", str(e))
                enrichedSigs.append(sig)
        self.signals = enrichedSigs

File: solution/go/enrichment/bad_domain.go
package main

import (
	"bufio"
	"context"
	"fmt"
	"log"
	"os"
	"strings"

	enpb "github.com/bearlyrunning/FindingTheNeedle/solution/go/generated/enrichmentpb"
	spb "github.com/bearlyrunning/FindingTheNeedle/solution/go/generated/signalpb"
)

const (
	indicatorPath = "../../data/indicators/bad_domain.csv"
	columnNum     = 2
)

func (bde *BadDomainEnricher) getName() string {
	return bde.name
}

func (bde *BadDomainEnricher) getSignals() []*spb.Signal {
	return bde.signals
}

func (bde *BadDomainEnricher) setSignals(sigs []*spb.Signal) {
	bde.signals = sigs
}

func loadIndicators() (map[string]enpb.Host_Platform, error) {
	hostToOS := make(map[string]enpb.Host_Platform)
	f, err := os.Open(indicatorPath)
	if err != nil {
		return nil, fmt.Errorf("failed to open file %s: %v", indicatorPath, err)
	}
	defer f.Close()
	s := bufio.NewScanner(f)
	for s.Scan() {
		strs := strings.Split(s.Text(), ",")
		if len(strs) == columnNum {
			ioc, p := strs[0], strs[1]
			hostToOS[ioc] = enpb.Host_Platform(enpb.Host_Platform_value[p])
		}
	}
	if err = s.Err(); err != nil {
		return nil, fmt.Errorf("failed to scan file: %v", err)
	}
	return hostToOS, nil
}

// Skipping unit test as this involves setting up mock which is out of scope for this workshop.
func (bde *BadDomainEnricher) enrich(ctx context.Context, c enpb.EnrichmentClient) {
	var enrichedSigs []*spb.Signal
	hostToOS, err := loadIndicators()
	if err != nil {
		log.Print(err)
	}
	for _, sig := range bde.signals {
		resp, err := c.IPToHost(ctx, &enpb.IP{Ip: sig.GetBadDomain().GetSourceIp()})
		if err != nil {
			log.Printf("failed IPToHost conversion, output signals for manual investigation: %v", err)
			enrichedSigs = append(enrichedSigs, sig)
		}
		if resp.GetPlatform() != hostToOS[sig.GetBadDomain().GetBadDomain()] {
			log.Printf("domain indicator %s observed from %s - an operating system that isn't impacted", sig.GetBadDomain().GetBadDomain(), resp.GetPlatform())
		} else {
			sig.GetBadDomain().Hostname = resp.GetName()
			enrichedSigs = append(enrichedSigs, sig)
		}
		continue
	}
	bde.signals = enrichedSigs
}
